pp.c

```diff
@@ -2753,24 +2738,23 @@ PP(pp_negate)
             /* It's publicly an integer */
         oops_its_an_int:
             if (SvIsUV(sv)) {
-                if (SvIVX(sv) == IV_MIN) {
-                    /* 2s complement assumption. */
-                    TARGi(SvIVX(sv), 1);/* special case: -((UV)IV_MAX+1) ==
-                                           IV_MIN */
-                    goto ret;
-                }
-                else if (SvUVX(sv) <= IV_MAX) {
-                    TARGi(-SvIVX(sv), 1);
+                if (SvUVX(sv) <= ABS_IV_MIN) {
+                    TARGi(NEGATE_2IV(SvUVX(sv)), 1);
                     goto ret;
                 }
             }
-            else if (SvIVX(sv) != IV_MIN) {
-                TARGi(-SvIVX(sv), 1);
+#ifdef PERL_PRESERVE_IVUV
+            else if (SvIVX(sv) < 0) {
+                TARGu(NEGATE_2UV(SvIVX(sv)), 1);
                 goto ret;
             }
-#ifdef PERL_PRESERVE_IVUV
             else {
-                TARGu((UV)IV_MIN, 1);
+                TARGi(-SvIVX(sv), 1);
+                goto ret;
+            }
```

Comment:
I'm not entirely convinced this change is equivalent to the old code, but then I'm not convinced the difference (if any) is observable from Perl code.

Reply:
Both old code and new code will return UV only if the argument is IV_MIN (note that `TARGu` actually will set `TARG` to IV if its argument is in the range of IV), so the any difference should not be observable from Perl code.